File: backend/core/spindle.py
import cv2 as cv
import numpy as np

from utils.debug import save_debug_image


def detect_spindle(
    image,
    refined_center,
    outer,
    search_offset=15,
    sample_count=180,
    sample_offset=3,
    debug=False
):
    """
    Fast + accurate spindle detector.

    Strategy:
    - local search around refined center
    - physically constrained radius
    - polarity agnostic
    - circular consistency scoring

    Returns:
    {
        "x": float,
        "y": float,
        "radius_px": float,
        "score": float
    }
    """

    if len(image.shape) == 3:

        gray = cv.cvtColor(
            image,
            cv.COLOR_BGR2GRAY
        )

    else:
        gray = image.copy()

    h, w = gray.shape[:2]

    cx0, cy0 = refined_center

    # -------------------------
    # expected spindle radius
    # -------------------------

    pixels_per_mm = (
        outer["radius_px"]
        / 152.4
    )

    expected_radius = (
        pixels_per_mm
        * 3.62
    )

    min_radius = max(
        3,
        int(expected_radius * 0.6)
    )

    max_radius = int(
        expected_radius * 1.4
    )

    # -------------------------
    # local normalize
    # suppress label texture
    # -------------------------

    # blur = cv.GaussianBlur(
    #     gray,
    #     (0, 0),
    #     sigmaX=5
    # )

    # normalized = cv.normalize(
    #     cv.absdiff(
    #         blur,
    #         gray
    #     ),
    #     None,
    #     0,
    #     255,
    #     cv.NORM_MINMAX
    # )

    normalized = gray

    # -------------------------
    # precompute trig
    # -------------------------

    angles = np.linspace(
        0,
        2 * np.pi,
        sample_count,
        endpoint=False
    )

    cosines = np.cos(
        angles
    ).astype(np.float32)

    sines = np.sin(
        angles
    ).astype(np.float32)

    best_score = -1e9
    best = None

    # -------------------------
    # local center search
    # -------------------------

    for dx in range(
        -search_offset,
        search_offset + 1
    ):

        for dy in range(
            -search_offset,
            search_offset + 1
        ):

            cx = cx0 + dx
            cy = cy0 + dy

            center_penalty = (
                np.hypot(dx, dy)
                * 0.15
            )

            for r in range(
                min_radius,
                max_radius + 1
            ):

                inner_r = (
                    r
                    - sample_offset
                )

                outer_r = (
                    r
                    + sample_offset
                )

                if inner_r <= 0:
                    continue

                # -------------------------
                # sample ring
                # -------------------------

                x1 = (
                    cx
                    + inner_r
                    * cosines
                ).astype(np.int32)

                y1 = (
                    cy
                    + inner_r
                    * sines
                ).astype(np.int32)

                x2 = (
                    cx
                    + outer_r
                    * cosines
                ).astype(np.int32)

                y2 = (
                    cy
                    + outer_r
                    * sines
                ).astype(np.int32)

                valid = (
                    (x1 >= 0)
                    & (x1 < w)
                    & (y1 >= 0)
                    & (y1 < h)
                    & (x2 >= 0)
                    & (x2 < w)
                    & (y2 >= 0)
                    & (y2 < h)
                )

                if (
                    np.count_nonzero(
                        valid
                    )
                    <
                    sample_count * 0.9
                ):
                    continue

                inner_values = normalized[
                    y1[valid],
                    x1[valid]
                ]

                outer_values = normalized[
                    y2[valid],
                    x2[valid]
                ]

                transition = (
                    np.mean(
                        outer_values
                    )
                    -
                    np.mean(
                        inner_values
                    )
                )

                # -------------------------
                # polarity agnostic
                # -------------------------

                strength = abs(
                    transition
                )

                radius_penalty = (
                    abs(
                        r
                        - expected_radius
                    )
                    * 0.35
                )

                # -------------------------
                # circular consistency
                # -------------------------

                consistency_penalty = (
                    np.std(
                        outer_values
                        -
                        inner_values
                    )
                    * 0.35
                )

                score = (
                    strength
                    - consistency_penalty
                    - center_penalty
                    - radius_penalty
                )

                if score > best_score:

                    # print(
                    #     "\nNEW BEST"
                    # )

                    # print(
                    #     "center:",
                    #     (cx, cy)
                    # )

                    # print(
                    #     "radius:",
                    #     r
                    # )

                    # print(
                    #     "transition:",
                    #     strength
                    # )

                    # print(
                    #     "consistency_penalty:",
                    #     consistency_penalty
                    # )

                    # print(
                    #     "center_penalty:",
                    #     center_penalty
                    # )

                    # print(
                    #     "radius_penalty:",
                    #     radius_penalty
                    # )

                    # print(
                    #     "score:",
                    #     score
                    # )

                    best_score = score

                    best = {
                        "x":
                            float(cx),

                        "y":
                            float(cy),

                        "radius_px":
                            float(r),

                        "score":
                            float(score)
                    }

    if best is None:

        raise RuntimeError(
            "Spindle detection failed"
        )

    debug_img = cv.cvtColor(
        gray.copy(),
        cv.COLOR_GRAY2BGR
    )

    cx = best["x"]
    cy = best["y"]
    r = best["radius_px"]

    angles_debug = np.linspace(
        0,
        2 * np.pi,
        72,
        endpoint=False
    )

    for theta in angles_debug:

        x1 = int(
            cx
            + (r - sample_offset)
            * np.cos(theta)
        )

        y1 = int(
            cy
            + (r - sample_offset)
            * np.sin(theta)
        )

        x2 = int(
            cx
            + (r + sample_offset)
            * np.cos(theta)
        )

        y2 = int(
            cy
            + (r + sample_offset)
            * np.sin(theta)
        )

        cv.line(
            debug_img,
            (x1, y1),
            (x2, y2),
            (0,255,255),
            1
        )

    cv.circle(
        debug_img,
        (
            int(cx),
            int(cy)
        ),
        int(r),
        (0,255,0),
        2
    )

    cv.circle(
        debug_img,
        (
            int(cx),
            int(cy)
        ),
        3,
        (0,0,255),
        -1
    )

    save_debug_image(debug, "08_spindle_debug.png", debug_img)

    return best

File: frontend/src/engine/pipeline/spindle.ts
// Mirrors backend/core/spindle.py → detect_spindle()
//
// Same ring-sampling hot loop as radialBoundary but with spindle-specific scoring:
//  - polarity-agnostic: strength = |mean(outer) - mean(inner)|
//  - radius_penalty = |r - expected_radius| * 0.35
//  - consistency_penalty = std(outer - inner) * 0.35
//
// Expected spindle radius: outer.radiusPx / 152.4 * 3.62 mm

import type { VinylGeometry, SpindleResult } from '../types';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function detectSpindle(
  cv: any,
  src: any,   // RGBA Mat (microRectified)
  refinedCenter: [number, number],
  outer: VinylGeometry,
  searchOffset = 15,
  sampleCount = 180,
  sampleOffset = 3,
): SpindleResult {
  // Convert RGBA → gray
  const gray = new cv.Mat();
  try {
    cv.cvtColor(src, gray, cv.COLOR_RGBA2GRAY);

    const pixels: Uint8Array = gray.data as Uint8Array;
    const step: number = (gray.step as Uint32Array)[0];
    const imgW = gray.cols;
    const imgH = gray.rows;

    const ppm = outer.radiusPx / 152.4;
    const expectedRadius = ppm * 3.62;
    const minRadius = Math.max(3, Math.round(expectedRadius * 0.6));
    const maxRadius = Math.round(expectedRadius * 1.4);

    const [cx0, cy0] = refinedCenter;

    // Pre-compute trig tables
    const cosines = new Float32Array(sampleCount);
    const sines   = new Float32Array(sampleCount);
    for (let i = 0; i < sampleCount; i++) {
      const a = (2 * Math.PI * i) / sampleCount;
      cosines[i] = Math.cos(a);
      sines[i]   = Math.sin(a);
    }

    const innerVals = new Float32Array(sampleCount);
    const outerVals = new Float32Array(sampleCount);

    let bestScore = -1e9;
    let bestCx = cx0, bestCy = cy0, bestR = minRadius;

    for (let dx = -searchOffset; dx <= searchOffset; dx++) {
      for (let dy = -searchOffset; dy <= searchOffset; dy++) {
        const cx = cx0 + dx;
        const cy = cy0 + dy;
        const centerPenalty = Math.hypot(dx, dy) * 0.15;

        for (let r = minRadius; r <= maxRadius; r++) {
          const innerR = r - sampleOffset;
          const outerR = r + sampleOffset;
          if (innerR <= 0) continue;

          let validCount = 0;
          for (let i = 0; i < sampleCount; i++) {
            const x1 = Math.round(cx + innerR * cosines[i]);
            const y1 = Math.round(cy + innerR * sines[i]);
            const x2 = Math.round(cx + outerR * cosines[i]);
            const y2 = Math.round(cy + outerR * sines[i]);
            if (x1 >= 0 && x1 < imgW && y1 >= 0 && y1 < imgH &&
                x2 >= 0 && x2 < imgW && y2 >= 0 && y2 < imgH) {
              innerVals[validCount] = pixels[y1 * step + x1];
              outerVals[validCount] = pixels[y2 * step + x2];
              validCount++;
            }
          }
          if (validCount < sampleCount * 0.9) continue;

          // mean(outer) - mean(inner)
          let innerSum = 0, outerSum = 0, diffSumSq = 0, diffSum = 0;
          for (let i = 0; i < validCount; i++) {
            innerSum += innerVals[i];
            outerSum += outerVals[i];
          }
          const innerMean = innerSum / validCount;
          const outerMean = outerSum / validCount;
          const transition = outerMean - innerMean;

          for (let i = 0; i < validCount; i++) {
            const d = outerVals[i] - innerVals[i];
            diffSum  += d;
            diffSumSq += d * d;
          }
          const diffMean = diffSum / validCount;
          const consistencyPenalty =
            Math.sqrt(Math.max(0, diffSumSq / validCount - diffMean * diffMean)) * 0.35;
          const radiusPenalty = Math.abs(r - expectedRadius) * 0.35;

          const score = Math.abs(transition) - consistencyPenalty - centerPenalty - radiusPenalty;

          if (score > bestScore) {
            bestScore = score;
            bestCx = cx; bestCy = cy; bestR = r;
          }
        }
      }
    }

    if (bestScore === -1e9) throw new Error('Spindle detection failed');

    return { x: bestCx, y: bestCy, radiusPx: bestR, score: bestScore };
  } finally {
    gray.delete();
  }
}
